pkg/endpoints/endpoints.go

```diff
@@ -19,19 +19,27 @@ type endpoint struct {
 }
 
 type endpointGroup struct {
+	mtx       sync.RWMutex
 	ports     map[string]int32
 	endpoints map[string]endpoint
-	active    *sync.Cond
-	mtx       sync.Mutex
+
+	active *sync.Cond
 }
 
-func (e *endpointGroup) getHost(portName string) string {
-	e.mtx.Lock()
-	defer e.mtx.Unlock()
-
-	for len(e.endpoints) == 0 {
-		e.active.Wait()
-	}
+// getBestHost returns the best host for the given port name. It blocks until there are available endpoints
+// in the endpoint group.
+//
+// It selects the host with the minimum in-flight requests among all the available endpoints.
+// The host is returned as a string in the format "IP:Port".
+//
+// Parameters:
+// - portName: The name of the port for which the best host needs to be determined.
+//
+// Returns:
+// - string: The best host with the minimum in-flight requests.
+func (e *endpointGroup) getBestHost(portName string) string {
```

Comment:
I think we are currently lacking the propagation of context cancellation down into this function where I think it needs to reach. Otherwise, this can hang forever (see comments in the timeout wrapper function). I am not sure you can accomplish this while utilizing the `sync.Cond` type. I think it might be best to heed the authors notes in the [sync.Cond docs](https://pkg.go.dev/sync#Cond) and find a solution that utilizes channels and respects context cancellation.